certification/internal/policy/operator/certified_images.go

```diff
@@ -0,0 +1,98 @@
+package operator
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/operator-framework/operator-manifest-tools/pkg/image"
+	"github.com/operator-framework/operator-manifest-tools/pkg/pullspec"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification/pyxis"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type imageFinder interface {
+	FindImagesByDigest(context.Context, []string) ([]pyxis.CertImage, error)
+}
+
+// CertifiedImagesCheck scans the CSV and validates that all refereenced images are certified.
+type certifiedImagesCheck struct {
+	ImageFinder        imageFinder
+	nonCertifiedImages []string
+}
+
+func NewCertifiedImagesCheck(imageFinder imageFinder) *certifiedImagesCheck {
+	return &certifiedImagesCheck{
+		ImageFinder: imageFinder,
+	}
+}
+
+func (p *certifiedImagesCheck) Validate(ctx context.Context, imgRef certification.ImageReference) (bool, error) {
+	imageDigests, err := p.dataToValidate(ctx, imgRef.ImageFSPath)
+	if err != nil {
+		return false, err
+	}
+
+	return p.validate(ctx, imageDigests)
+}
+
+func (p *certifiedImagesCheck) dataToValidate(ctx context.Context, imagePath string) ([]string, error) {
+	operatorManifests, err := pullspec.FromDirectory(imagePath, pullspec.DefaultHeuristic)
+	if err != nil {
+		return nil, err
+	}
+	imageNames, err := image.Extract(operatorManifests)
+	if err != nil {
+		return nil, err
+	}
+
+	imageDigests := make([]string, 0, len(imageNames))
+	for _, img := range imageNames {
+		digest, err := name.NewDigest(img)
+		if err != nil {
+			log.Warningf("Image does not appear to be pinned: %s: %v", img, err)
+			p.nonCertifiedImages = append(p.nonCertifiedImages, img)
+			continue
+		}
+		imageDigests = append(imageDigests, digest.DigestStr())
+	}
+
+	return imageDigests, nil
+}
+
+func (p *certifiedImagesCheck) validate(ctx context.Context, imageDigests []string) (bool, error) {
+	pyxisImages, err := p.ImageFinder.FindImagesByDigest(ctx, imageDigests)
+	if err != nil {
+		return false, err
+	}
+
+	for _, img := range pyxisImages {
+		if !img.Certified {
+			fullImg := fmt.Sprintf("%s/%s@%s", img.Repositories[0].Registry, img.Repositories[0].Repository, img.DockerImageDigest)
+			p.nonCertifiedImages = append(p.nonCertifiedImages, fullImg)
+		}
+	}
+	return true, nil
+}
+
+func (p *certifiedImagesCheck) Name() string {
+	return "CertifiedImages"
```

Comment:
I think this needs to be more descriptive. "BundleImageReferencesAreCertified", "AllImagesAreCertified" or something to that extent.

Reply:
Done.